### PR TITLE
Create interface for generating JitBuilder types from C/C++ types

### PR DESCRIPTION
**Pull request by Leonardo2718**
This adds an interface to JitBuilder that allows instances of `TR::IlType` to be generated from C/C++ types.

Issue: #445

### CODE REVIEW

File: compiler/ilgen/TypeDictionary.hpp [hunk 1]
```diff
@@ -229,6 +229,176 @@ class TypeDictionary
...
+    * For integer and floating point types, the type traits library is used to
+    * identify the "family" of the specified type. For example, `std::is_integral<>`
+    * is used to identify all the integer types. The `sizeof` operator is then used
+    * to determine the size of the specified type. The combination of the type's
+    * family and size is used to select the function that gets selected.
+    * 
+    * For types that do not belong to a family (e.g. `void`), the size is not needed.
+    * 
+    * For pointer types, `std::remove_pointer<>` is used to get the type being
+    * pointed to. Iff `is_supported<>::value` evaluates to true for this type,
```
Comment (fjeremic) [inline compiler/ilgen/TypeDictionary.hpp]: Typo: "Iff", unless you meant "If and only if" in which case the sentence is slightly confusing.

Comment (Leonardo2718) [inline compiler/ilgen/TypeDictionary.hpp]: I did mean "if and only if", but I see how that makes it bit confusing. Thanks for pointing that out!

**Comment by fjeremic** [comment]
LGTM. Excellent documentation! I wonder how we would be able to support arrays of types.

**Comment by andrewcraik** [comment]
The thing that confused me to start with about this implementation is that you can generate a void* but a <struct>* is not supported. It comes down to if we support the type we expect to be able to dereference to something we know about and deref of the void* is essentially the user telling us to trust them and they know what they are doing. I'm not sure if this point should be made more explicit in the documentation.

Otherwise I agree it seems sane and sensible and a fine starting point.

**Comment by mstoodle** [comment]
Very nice job, @Leonardo2718 ! Comprehensive tests, good documentation, well written. Thanks!!

**Pull request merged**